app/lib/expression_viz_service.rb

```diff
@@ -137,6 +137,44 @@ def self.load_expression_data_array_points(study, gene, cluster, annotation, sub
     viz_data
   end
 
+  def self.load_correlated_data_array_scatter(study, genes, cluster, annotation,  subsample_threshold=nil)
+    viz_data = ClusterVizService.load_cluster_group_data_array_points(study, cluster, annotation, subsample_threshold=nil, include_coords: false)
+
+    gene0_expression = viz_data[:cells].map { |cell| genes[0]['scores'][cell].to_f.round(4) }
+    gene1_expression = viz_data[:cells].map { |cell| genes[1]['scores'][cell].to_f.round(4) }
+
+    viz_data[:x] = gene0_expression
+    viz_data[:y] = gene1_expression
+    viz_data[:pearsonsR] = pearsons_r(gene0_expression, gene1_expression).to_f.round(4)
+    viz_data
+  end
+
+  # adapted from https://blog.chrislowis.co.uk/2008/11/24/ruby-gsl-pearson.html
+  # note that the age of the blog means his performance numbers of the ruby version are off
+  # in local testing, this can compute correlation for 1MM cells in 0.5 seconds
+  def self.pearsons_r(x,y)
```

Comment:
Style: I get this was copied from the above source, but it needs to be reformatted for readability and general Ruby style.  The variables are camel-case, and the spacing is all off in places.

Reply:
heh, yeah, I cleaned it up a little bit, but you're right it needs more work